### DISCUSSION TITLE
e32-400m30s (sx1278+PA) inverted SetOutputPower result

### DISCUSSION THREAD

**Discussion by whoim2** [jgromes/RadioLib#813] [Q&A] [answered]
Strange SetOuptupPower behavior for e32-400m30s module (sx1278 + unknown PA). As part of the experiment, I set the power:
SetOutputPower(2), (6) and (12). At 2 I get a burning module, at 6 moderately hot, at 12 almost cold.
When transmitting, I set the TX_EN pin of the module and remove it when I receive a DIO0 interrupt.
Why can observational results be inverted relative to expectations?

**Comment by jgromes** [comment]
I have tested and measured the output from the SX127x using RF power meter and I'm quite confident that the output power matches the configuration.

The expectation that hot module = lots of power output is incorrect, it just means that most of the power is being reflected back into your module (or something is shorted) and the SX1278 over-current protection might be kicking in at higher power. 

Couple of things to investigate:

1. Check the connection to antenna, if the module is getting hot it would suggest most of the power is reflecting back into it instead of radiating from the antenna.
2. Does the SX127x use RFO or PA_BOOST to connect the RF output to the PA? RadioLib assumes PA_BOOST by default.
3. If you don't know what external PA is used, the module is useless to you since the PA will be frequency-dependent and will require some minimum/maximum input power level it can work with.

**Comment by whoim2** [comment]
Thanks for such a detailed answer!
Indeed, the rssi readings behave as expected, it is more stable at higher output power, and at the same time lower heating. I could not find information about the PA used and will try to find out the maximum possible gain value experimentally.
The default mode is used, which means PA Boost. I could not find information, what is the difference between them, if you tell me, I will be grateful. All I do apart from the library is enable 1 on the txen rxen pins of the module.
I will also ask the manufacturer about PA, thank you!

**Comment by jgromes** [comment]
> I could not find information, what is the difference between them

Let's take a look in the SX1278 datasheet. On the first page, you will find the block diagram of the device. On it , you will find both PA_BOOST as well as RFO (I marked them in red in the picture below).

![Screenshot_65](https://github.com/jgromes/RadioLib/assets/3407192/f2ab5091-4a03-49ee-95eb-32f2ec85eb70)

As you can see, those are two completely different pins - you can check the pinout of the SX1278 further in the datasheet. So if you use the wrong one, it might be unconnected on your module, and if you transmit from it, almost all of the power will be reflected back into the SX1278 damaging it.

On a side note, this doesn't seem like an issue in the library so far, so I will convert this into a discussion.

**Comment by whoim2** [comment]
[manual for this module](https://www.ebyte.com/en/downpdf.aspx?id=1624)

I see the following in the description:

> >NRST, TXEN and RXEN pins must be connected, while NRST controls the reset of the chip, TXENandRXENpins control the RF switch;
> When transmitting, set DIO_7 pin to high level and DIO_13, DIO_14 pins to low level; When receiving, set DIO_7 pin to low level and DIO_13, DIO_14 pins to high level;

this will not help determine if PA_BOOST or RFO?

**Comment by whoim2** [comment]
Im find picture of module.. say me please, 27 pjn of sx1278 its pa_boost, this pin used?
![Screenshot_20230824-042541](https://github.com/jgromes/RadioLib/assets/13419599/05e09411-6a90-48d0-818f-d0bf74413333)
![Screenshot_20230824-042554](https://github.com/jgromes/RadioLib/assets/13419599/6be4a5d0-5008-4ce0-bba4-0c14537230c5)

**Reply by jgromes** [reply, marked as answer]
Looks like the module is using PA_BOOST, you can clearly see that it is connected to the impedance matching network:

![Screenshot_66](https://github.com/jgromes/RadioLib/assets/3407192/8ea63059-0e91-4835-8ac4-6ec6869647a6)

**Reply by whoim2** [reply]
Thanks! Now, my max output power is 30dbm - pa gain (unknown).
Im try all values and check current power.
Ant have problems, im check it.

**Comment by whoim2** [comment]
I can't figure out exactly if my code is supposed to do this (written in the module developer's guide), or is the developer just warning that these pins are being used? If I must, then I cannot find for sx1278 something like setDio()
![photo_2023-08-31_16-00-32](https://github.com/jgromes/RadioLib/assets/13419599/7f353547-144a-47ff-9c50-5ad6bae362e8)

**Reply by jgromes** [reply]
Since the SX1278 only has DIO0 - DIO5, I'm guessing the manual actually means the module pins 7, 13 and 14. However, pin 13 is marked as not connected and 14 as connected to SX1278 DIO3. Doesn't really make sense to me, you will have to contact the manufacturer for clarification.

I would guess that to transmit, you will need to pull TXEN high and RXEN low, and to receive, vice versa. RadioLib can do that for you automatically if you call `radio.setRfSwitchPins(rxEn, txEn)`.

**Comment by whoim2** [comment]
With the manufacturer, everything is complicated. I could do screenshots of the correspondence where they check that their manual has everything you need to read the manual for the sx1278 chip, and that they cannot tell me which PA is installed or at least its power, input restrictions and what method it is connected to (PA_BOOST or RFO). We have dealt with pin. I really don't want to open the module to read the markings. Yes, and most likely it was destroyed.
I am switching txen/rxen on my own, is this acceptable?
Before transmit I call txen(), and in interrupt immediately after receiving Dio0 signal rxen()

![Screenshot_1](https://github.com/jgromes/RadioLib/assets/13419599/3b7f38aa-cafd-4b67-8e36-2d7480101108)

**Reply by jgromes** [reply]
> With the manufacturer, everything is complicated

Sorry to hear that, unfortunately this is very common with Chinese manufacturers ...

> I am switching txen/rxen on my own, is this acceptable?

If you do it before you start transmitting/receiving then it doesn't matter. However, from your decription, for reception you should be calling it before `startReceive` or `receive`, DIO is activated when the packet is fully received, which is too late.

**Comment by whoim2** [comment]
im replace my txen/rxen switch to radio.setRfSwitchPins, no changes.
At setOutputPower(17) best rssi and modules cold, at setOutputPower(3) rssi a little lower but the modules begin to warm up noticeably, almost to the point of burning. I checked the antennas, agreed on a common frequency where the SWR is less than 1.5 (even 1.1) and use this range (I have a floating frequency on each packet)
UDP there is a feeling that the overall heating has become a little less, although the overall dependence remains the same

**Reply by jgromes** [reply]
Not much I can do from here unfortunately. It would be interesting to see the power at the output if you happen to have an RF power meter.

**Comment by whoim2** [comment]
No, I called txen before the transfer. After the end of the transfer, I called rxen on the dio0 signal so as not to heat the module in vain.
Unfortunately, there is no power meter :(
Once I assembled a field detector on a microammeter and an RF diode, but I can’t find it.
Maybe I'll order an inexpensive meter.
